pycrostates/cluster/kmeans.py

```diff
@@ -0,0 +1,289 @@
+        # all channels recorded the same value at the same time (=0 due to
+        # average reference)
+        # ---------------------------------------------------------------------
+        data = data[:, np.linalg.norm(data.T, axis=1) != 0]
+        n_channels, n_samples = data.shape
+        data_sum_sq = np.sum(data ** 2)
+
+        # Select random time points for our initial topographic maps
+        init_times = random_state.choice(
+            n_samples, size=n_clusters, replace=False)
+        maps = data[:, init_times].T
+        # Normalize the maps
+        maps /= np.linalg.norm(maps, axis=1, keepdims=True)
+
+        prev_residual = np.inf
+        for _ in range(max_iter):
+            # Assign each sample to the best matching microstate
+            activation = maps.dot(data)
+            segmentation = np.argmax(np.abs(activation), axis=0)
+
+            # Recompute the topographic maps of the microstates, based on the
+            # samples that were assigned to each state.
+            for state in range(n_clusters):
+                idx = (segmentation == state)
+                if np.sum(idx) == 0:
+                    maps[state] = 0
+                    continue
+
+                # Find largest eigenvector
+                maps[state] = data[:, idx].dot(activation[state, idx])
+                maps[state] /= np.linalg.norm(maps[state])
+
+            # Estimate residual noise
+            act_sum_sq = np.sum(
+                np.sum(maps[segmentation].T * data, axis=0) ** 2)
+            residual = abs(data_sum_sq - act_sum_sq)
+            residual /= float(n_samples * (n_channels - 1))
+
+            # check convergence
+            if (prev_residual - residual) < (tol * residual):
+                break
+
+            prev_residual = residual
+
+        else:
+            logger.warning(
+                'K-means algorithm failed to converge. Please adapt the '
+                'tolerance and the maximum number of iteration.')
+            converged = False
+        converged = True
+
+        return maps, converged
+
+    # --------------------------------------------------------------------
+    @property
+    def n_init(self):
+        """
+        Number of time the k-means algorithm is run with different centroid
+        seeds.
+
+        :type: `int`
+        """
+        return self._n_init
+
+    @property
+    def max_iter(self):
+        """
+        Number of maximum iterations of the k-means algorithm for a single run.
+
+        :type: `int`
+        """
+        return self._max_iter
+
+    @property
+    def tol(self):
+        """
+        Relative tolerance.
+
+        :type: `float`
+        """
+        return self._tol
+
+    @property
+    def random_state(self):
+        """
+        Random state.
+
+        :type: `None` | `int` | `~numpy.random.RandomState`
+        """
+        return self._random_state
+
+    @property
+    def GEV(self):
```

Comment:
Maybe we can follow the sklearn convention for this one  -> Attributes that have been estimated from the data must always have a name ending with trailing underscore
def GEV_(self):

and change self._GEV to self.GEV_

Reply:
Ok, so I'll keep the attribute as `self._gev` (private) but rename the property to `GEV_`

Reply:
To be sure, there are 3 variables that have been estimated from data and should end with a trailing underscore: `GEV`, `labels` and `cluster_centers`. Correct?